napari/components/_dims/model.py

```diff
@@ -1,143 +1,255 @@
-        self.indices[axis] = value
-        self._need_redraw = True
-        self._update()
-
-    def _calc_max_dims(self):
-        """Calculates the number of maximum dimensions in the contained images.
-        """
-        max_dims = 0
+        slice_list = []
+        project_list = []
+        z = zip(self.mode, self.display, self.point, self.interval, self.range)
+        for (mode, display, point, interval, range) in z:
+            if mode == DimsMode.POINT or mode is None:
+                if display:
+                    # no slicing, cropping or projection:
+                    project_list.append(False)
+                    slice_list.append(slice(None, None, None))
+                else:
+                    # slice:
+                    project_list.append(False)
+                    slice_list.append(int(round(point)))
+            elif mode == DimsMode.INTERVAL:
+                if display:
+                    # crop for display:
+                    project_list.append(False)
+                    if interval is None:
+                        slice_list.append(slice(None))
+                    else:
+                        slice_list.append(slice(int(round(interval[0])),
+                                          int(round(interval[1]))))
+                else:
+                    # crop before project:
+                    project_list.append(True)
+                    if interval is None:
+                        slice_list.append(slice(None))
+                    else:
+                        slice_list.append(slice(int(round(interval[0])),
+                                          int(round(interval[1]))))
+
+        slice_tuple = tuple(slice_list)
+        project_tuple = tuple(project_list)
+
+        return slice_tuple, project_tuple
 
-        for layer in self.viewer.layers:
-            dims = layer.ndim
-            if dims > max_dims:
-                max_dims = dims
+    @property
+    def indices(self):
+        """
+        Indices for slicing
 
-        self._max_dims = max_dims
+        Returns
+        -------
+        slice : tuple
+            The slice tuple
+        """
+        return self.slice_and_project[0]
 
-        self._need_slider_update = True
-        self._update()
+    def set_all_ranges(self, all_ranges: Sequence[Union[int, float]]):
+        """Sets ranges for all dimensions
 
-    def _calc_max_shape(self):
-        """Calculates the maximum shape of the contained images.
+        Parameters
+        ----------
+        ranges : tuple
+            Ranges of all dimensions
         """
-        shapes = (layer.shape for layer in self.viewer.layers)
-        self._max_shape = _compute_max_shape(shapes, self.max_dims)
-
-    def _on_layers_change(self, event):
-        """Called whenever a layer is changed.
+        self.ndim = len(all_ranges)
+        self._set_2d_viewing()
+        self.range = all_ranges
+
+    def set_range(self, axis: int, range: Sequence[Union[int, float]]):
```

Comment:
this implies that you're setting the `range` attribute. to improve clarity, I suggest making the attribute name plural or private

Reply:
i have made these all properties that return copies